java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java

```diff
@@ -156,11 +157,21 @@ synchronized ChangedSources getAndResetChangedSources(Label label) {
 
     switch (data.state) {
       case WAITING_FOR_SOURCES:
+        if (!data.changedNonCompilableSources.isEmpty()) {
```

Comment:
Why does this happen only in waiting for sources and not in collecting?

Reply:
This is due to the specific case, where proto files were edited, but the state is still "WAITING_FOR_SOURCES" which means, that the edited files could not be compared to sources, and full compile happens "just in case". If the state is COLLECTING, it means, that the edited files are being checked against the sources, so there weren't any changed proto files that were in the sources. If there were changed files that were in the sources , the state would be set to NON_COMPILABLE_CHANGES, and full compile would happen due to that. 

If not for this specific case, the code would be much more simple, as there would not be need for data.changedNonCompilableSources property at all :/